Multi descriptor: Decouple transaction building logic

evanlinjin opened the pull request:
## Description

Closes bitcoindevkit/bdk_wallet#188

The end goal of this PR is to have the ability to a single wallet with multiple descriptors.

## Notes to the reviewers

In the Discord channel @ `#multi-descriptor-wallets`, @afilini [suggested a roadmap](https://discord.com/channels/753336465005608961/991846404068749405/993812164685082655):

> I think a reasonable roadmap would be:
> 1. pr to add "database factories" or something similar
> 2. pr to add "multi-wallet" support (collection of the current Wallet struct). The subwallets will only use the "external" keychain, and at the multi-wallet level you can define what to use as internal/external
> 3. pr to remove the two keychains from the wallet struct, deal with renaming and deprecating stuff (users should switch from Wallet to MultiWallet with two subwallets etc)

The "database factories" structure is already implemented in a separate PR: bitcoindevkit/bdk#654 (And can be used later).

This PR mainly addresses the initial portion of step 2.

### My Approach

The most difficult aspect of adding multi-descriptor capability is in the transaction creation. Everything else from retrieving address, UTXOs, signing, transactions, balances and syncing are trivial tasks. Hence, my approach is to first decouple the transaction building logic from `Wallet` so that it can be reused for multi-descriptor (or even for users of BDK that don't want a wallet). In that being said, I have kept what @afilini said in mind (paraphrasing)...

* Avoid resulting in ***a huge pr that breaks everything***.
* ... ***I'm sure it's gonna take a while to get pr 2 to work well and it's probably better if we avoid replacing the current `Wallet` with it immediately.***

Therefore, I have ensured the following things for my approach:

1. Instead of moving logic out of `Wallet`, I have turned various member functions into static functions (renamed to start with `_`). The original functions will call these static functions. This way, we do not change the "external interface" of `Wallet` and more-or-less keep these bits and pieces in the same place (so hopefully git will not complain so much when we merge).

2. I have introduced a new trait `SpendableCollection` to replace the functionality of `Wallet`/`Database` within `TxBuilder`. To keep `Wallet` working with the new modified `TxBuilder`, a struct `SpendablesFromDb` is formed which implements `SpendableCollection`, and wraps `Database` (and some fields of `Wallet`).

### Noteworthy changes

Not in any particular order...

* Introduction of the `SpendableCollection` trait (and `SpendableCollectionInner` which defines it's "explicitly required" methods). These traits represent a collection of owned and spendable UTXO's, Descriptors and their associated transactions. Note that it's members `SpendableCollectionInner::get_{external|internal}_descriptors` returns `Vec`s (so we can support multiple descriptor implementations in the very near future).
* The `CoinSelectionAlgorithm` trait's type constraint has been moved to it's function definition (the constrained type was only used as input anyway) - I hope this change is okay?
* Instead of referencing descriptors by `KeychainKind`, I have changed to use the `ExtendedDescriptor` directly (again, for muti-descriptors in the future). This change has some small caveats...
  * https://github.com/bitcoindevkit/bdk/blob/574dcc4ee3f045cf5a3a0cc61667b5aeff370be5/src/wallet/spendable_collection.rs#L73-L101
  * https://github.com/bitcoindevkit/bdk/blob/574dcc4ee3f045cf5a3a0cc61667b5aeff370be5/src/wallet/mod.rs#L5279-L5283
* `TxBuilder`, other than needing the `BatchDatabase` constraint with `SpendableCollection`, uses static members from `Wallet` (as mentioned above). Hence, we needed to add another phantom constraint imposed by the constraints of those `Wallet` methods (this can be removed once we ACTUALLY move stuff out of `Wallet`).
* TODO @evanlinjin: I have probably forgotten to mention something here... I will keep this space updated.

### Tests

It seems that I have not broken the functionality of `Wallet` and all the non-doc tests continue to pass.

### Next steps

Once @afilini and everyone else is happy with my approach thus far, I will create a multi-descriptor implementation of `SpendableCollection` and do tests to see if multi-descriptor transaction creation works as expected.

Since `CoinSelectionAlgorithm` now takes a `SpendableCollection` as an input (instead of a `BatchDatabase`, I need to ensure that `SpendableCollection` is satisfactory for all coin selection algorithms.

I also still need to decouple Fee-bumping logic.

## Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [ ] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

evanlinjin committed:
.gitignore for vscode

evanlinjin committed:
Change `TxParams` for multi-descriptor context

Store policy paths in a map with ExtendedDescriptor as key. This way we
can re-use the TxParams struct in multi-descriptor/multi-wallet
contexts.

evanlinjin committed:
Modify `Error::SpendingPolicyRequired`

Use `ExtendedDescriptor` instead of `Keychain`

evanlinjin committed:
WIP: Decouple `TxBuilder` from `Wallet`.

This is a WIP.

Explanation for my approach (will be / is) provided in the PR's
description box.

evanlinjin committed:
WIP: Decoupled signing logic

evanlinjin committed:
WIP: Added TxBuilder::finish_and_sign

evanlinjin committed:
WIP: Decouple fee-bump logic

evanlinjin committed:
WIP: Change SpendableCollection to use boxed Iterators

evanlinjin committed:
WIP: Rename `SpendableCollection` to `MultiTracker` and tweaks.

* Added `MultiTracker::latest_blockheight` method.
* Update documentation to reflect changes.

evanlinjin committed:
WIP: Introduce LocalOutputItem

afilini commented on src/wallet/tx_builder.rs:
```diff
@@ -117,23 +123,24 @@ impl TxBuilderContext for BumpFee {}
-    pub(crate) internal_policy_path: Option<BTreeMap<String, Vec<usize>>>,
-    pub(crate) external_policy_path: Option<BTreeMap<String, Vec<usize>>>,
+    pub(crate) descriptor_policy_paths:
+        BTreeMap<&'a ExtendedDescriptor, BTreeMap<String, Vec<usize>>>,
```
Maybe we can avoid the lifetime by using something else as a key? like the descriptor's checksum?

evanlinjin replied:
I will change to use checksum once https://github.com/bitcoindevkit/bdk/pull/671 is approved and merged :)

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+///! MultiTracker
```
This will eventually need the usual header that we have in all the other files

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+    /// * Confirmed height.
+    /// * Descriptor + Child index.
+    /// * Satisfaction weight of scriptPubKey (weight of corresponding scriptSig/witness data).
+    fn iter_unspent(&self) -> Result<Box<dyn Iterator<Item = (LocalUtxo, usize)> + '_>, Error>;
```
We actually have a `WeightedUtxo` struct that is basically a utxo + the weight, maybe that's more convenient here

evanlinjin replied:
However, `WeightedUtxo` can be of a foreign `Utxo` so it is not as "direct" as using an union here (since this method is about iterating unspent, and owned UTXOs). Should I keep it as is?

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+
+        let secp = SecpCtx::new();
+
+        for index in 0..2100_u32 {
```
Wait, that's a lot! Deriving all the descriptors up to index 2100 may take a very long time. Do you think it makes sense to add an extra argument for a "max index"? Most of the time I guess the caller would only need to derive up to 100/200

evanlinjin replied:
This is the default implementation of the method and I intended it just to be there for convenience. But you are right, so I've changed it to check for 200 iterations, and added a comment saying that this method should be overloaded. Would that be enough? It seems inelegant to add another input imo.

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+                    .raw
+                    .output
+                    .get(outpoint.vout as usize)
+                    .map(Clone::clone))
```
Should we return an error if a tx doesn't have the requested vout? I kinda feel like these are swapped around, we should return `None` if we don't know about the outpoint, but if we know the tx and the vout is missing, then it's clearly invalid

evanlinjin replied:
Resolved in 7649de62edc4610107c852185e87780a05a92ee6

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+
+        internal.append(&mut external);
+
+        let item = internal.iter().chain(external.iter()).next().unwrap();
```
I think here you are "appending" twice. First directly to the vector, and then when iterating you chain with external.

`append` probably empties the `external` vector, so in the end the code works fine but it looks weird.

evanlinjin replied:
Addressed in 400dbd19d9ae4c094da74213cb22ec1eef8a2e69

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+                let (item, _) = self
+                    // @evanlinjin: Will panic with default implementation on timeout.
+                    .get_path_of_script_pubkey(&utxo.txout.script_pubkey)
+                    .unwrap()?;
```
Oh is that a question mark on an option? I thought they were still unstable, it's awesome we can finally use them 😍

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+        // TODO: Add as struct field
+        let desc_map = self
+            .iter_descriptors()?
+            .map(|desc| (desc.descriptor.to_string(), desc))
```
I think here again you can just use the checksum which is shorter

evanlinjin replied:
Will wait on https://github.com/bitcoindevkit/bdk/pull/671 :)

afilini commented on src/wallet/multi_tracker.rs:
```diff
@@ -0,0 +1,405 @@
+
+/// Implements [`MultiTracker`] with one external descriptor and an optional internal descriptor.
+// #[derive(Clone)]
+pub struct LegacyTracker<'a, D> {
```
Is there a specific reason why you chose to not let this struct own the descriptors and the database and instead use references?

afilini replied:
Is it to make this easier to clone for example? I see that in the `TxBuilder` you have to clone this, in that case it makes sense, but then the question is: why can't you store a reference in the `TxBuilder` like we used to do before with the `Wallet`?

evanlinjin replied:
@afilini It is just to avoid copying data in memory. The assumption is that one would only use `LegacyTracker` with the existence of `Wallet` (which is a good assumption imo). Since, the whole point of `LegacyTracker` is to get `Wallet` to "implement" `MultiTracker`, since transaction-building and signing logic is to be wrapped around `MultiTracker`. Right now this tx logic uses static functions of `Wallet` (which is confusing), but will avoid annoying git as much as possible. Those static functions, responsible for tx logic, should eventually be moved out of `Wallet` when the time is right.

afilini commented on src/wallet/coin_selection.rs:
```diff
@@ -144,7 +145,7 @@ impl CoinSelectionResult {
+pub trait CoinSelectionAlgorithm: std::fmt::Debug {
     /// Perform the coin selection
     ///
     /// - `database`: a reference to the wallet's database that can be used to lookup additional
```
The documentation will have to be updated eventually, just writing this as a reminder

evanlinjin replied:
Addressed in 2a302d2c7ef344331a23ee094b4bc52b106297c7

afilini commented on src/wallet/coin_selection.rs:
```diff
@@ -144,7 +145,7 @@ impl CoinSelectionResult {
 ///
 /// For an example see [this module](crate::wallet::coin_selection)'s documentation.
-pub trait CoinSelectionAlgorithm<D: Database>: std::fmt::Debug {
+pub trait CoinSelectionAlgorithm: std::fmt::Debug {
```
I'm trying to figure out how these changes affect my original idea for the coin selection: the reason why the `D` generic was defined here was that we wanted to offer a way to use custom databases in a custom coin selection without being restricted on the API in any way.

For example, let's say I have my custom database that exposes some kind of metric that I'd like to use for the coin selection, then I can implement `CoinSelectionAlgorithm<MyCustomDb>` and this would make the cs get a `&MyCustomDb` (not a generic `&dyn Database`), which would allow me to use any custom API I have.

It's not the end of the world to lose this ability because again it was something that probably nobody used, but I think it was kind of cool and I'm wondering if we can somehow preserve this even with the `MultiTracker`.

evanlinjin replied:
But since CoinSelectionAlgorithm is a trait, we can have a struct implement CoinSelectionAlgorithm (with special databases and parameters as fields) so we can still do those custom algorithms without the need for the generic type. Let me know if I've missed something though

afilini commented on src/wallet/coin_selection.rs:
```diff
@@ -649,21 +675,52 @@ mod test {
+    fn tx(utxo: &mut WeightedUtxo, lock_time: u32) -> (Transaction, Txid) {
+        let local = match &mut utxo.utxo {
+            Utxo::Local(local) => local,
+            _ => panic!(),
```
We may want to at least provide a good error message here

evanlinjin replied:
Addressed in 7432b07ad36942de3884d35906748d79c7e76183

afilini reviewed (commented):
I still feel like I'm trying to wrap my head around this, but conceptually I think it makes sense and I really like how this allows us to slowly move away from `Wallet` by making it "backward compatible" with `MultiTracker`.

evanlinjin committed:
WIP: Add header comments for multi_tracker.rs

evanlinjin committed:
WIP: Reduce iterations for MultiTracker:get_path_of_script_pubkey and comments

evanlinjin committed:
WIP: Fix `MultiTracker::get_output` logic

It should return `Ok(None)` when tx is not found, but return `Err` when
`OutPoint.vout` is invalid.

evanlinjin committed:
WIP: Fix `MultiTracker::new_auto_address` logic

evanlinjin committed:
WIP: Fix documentation in `coin_selection.rs`

evanlinjin committed:
WIP: Provide error message for `coin_selection::test::tx`

evanlinjin commented:
This work will be replaced with BDK Core/Chain.
